system/src/firmware_update.cpp

```diff
@@ -0,0 +1,432 @@
+
+FirmwareUpdate* FirmwareUpdate::instance() {
+    static FirmwareUpdate instance;
+    return &instance;
+}
+
+#if HAL_PLATFORM_RESUMABLE_OTA
+
+int FirmwareUpdate::initTransferState(size_t fileSize, const char* fileHash) {
+    std::unique_ptr<detail::TransferState> state(new(std::nothrow) detail::TransferState());
+    if (!state) {
+        return SYSTEM_ERROR_NO_MEMORY;
+    }
+    CHECK(state->partialHash.init());
+    CHECK(state->tempHash.init());
+    bool resumeTransfer = false;
+    const auto persist = &state->persist;
+    const int r = state->file.load(persist, sizeof(PersistentTransferState));
+    if (r == sizeof(PersistentTransferState)) {
+        if (persist->fileSize == fileSize && persist->partialSize <= fileSize &&
+                memcmp(persist->fileHash, fileHash, Sha256::HASH_SIZE) == 0) {
+            // Compute the hash of the partially transferred data stored in the OTA section
+            CHECK(state->partialHash.start());
+            char buf[OTA_FLASH_READ_BLOCK_SIZE] = {};
+            uintptr_t addr = HAL_OTA_FlashAddress();
+            const uintptr_t endAddr = addr + persist->partialSize;
+            while (addr < endAddr) {
+                const size_t n = std::min(endAddr - addr, sizeof(buf));
+                CHECK(HAL_OTA_Flash_Read(addr, (uint8_t*)buf, n));
+                CHECK(state->partialHash.update(buf, n));
+                addr += n;
+            }
+            CHECK(state->tempHash.copyFrom(state->partialHash));
+            CHECK(state->tempHash.finish(buf));
+            if (memcmp(persist->partialHash, buf, Sha256::HASH_SIZE) == 0) {
+                resumeTransfer = true;
+            }
+        }
+    } else if (r < 0 && r != SYSTEM_ERROR_NOT_FOUND && r != SYSTEM_ERROR_BAD_DATA) {
+        return r;
+    }
+    if (!resumeTransfer) {
+        state->file.clear();
+        memcpy(persist->fileHash, fileHash, Sha256::HASH_SIZE);
+        memset(persist->partialHash, 0, Sha256::HASH_SIZE);
+        persist->fileSize = fileSize;
+        persist->partialSize = 0;
+        CHECK(state->partialHash.start()); // Reset SHA-256 context
+    }
+    state->startOffset = persist->partialSize;
+    transferState_ = std::move(state);
+    return 0;
+}
+
+int FirmwareUpdate::updateTransferState(const char* chunkData, size_t chunkSize, size_t chunkOffset, size_t partialSize) {
+    const auto state = transferState_.get();
+    if (!state) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    const auto persist = &state->persist;
+    const size_t partialSizeBefore = persist->partialSize;
+    // Check if the chunk is adjacent to or overlaps with the contiguous fragment of the data for
+    // which we have already calculated the checksum
+    if (persist->partialSize >= chunkOffset && persist->partialSize < chunkOffset + chunkSize) {
+        const auto n = chunkOffset + chunkSize - persist->partialSize;
+        CHECK(state->partialHash.update(chunkData + persist->partialSize - chunkOffset, n));
```

Comment:
Yeah, the caller is not expected to overwrite the already stored data. We also erase the flash only once before an update so overwriting is not even technically possible in the current code. I'll update the comment above so that it's less confusing :+1:

Reply:
Updated.